### ISSUE TITLE
About 30 resamples indices

### ISSUE THREAD

**Issue by firrice** [angus924/quant#1]
Hi, i'm doing experiments on UCR, and would you mind offering resample indices for datasets? More specifically,  "indices" which  means the indices of contained series for every resample id in UCR.  I have tried aeon's stratified_resample_data() to generate indices, but they are different from the resample indices of some published paper, eg: 《MultiRocket》. And I found you have did experiments on UCR148，so i wish you offering the resample indices, and that will really do me a great help!

**Comment by firrice** [comment]
I also have tried tsml-java tools for generating 30 resamples(reference to: https://github.com/time-series-machine-learning/tsml-java/commit/057e9a0b1de70989ffee08e2def8f442fde7982d). But it is still different with aeon'result or indices files from MultiRocket's github :(

**Comment by angus924** [comment]
Hi @firrice,

Sorry for the slow reply.

This is a very good question.

Unfortunately, the resamples situation is just a bit of a mess.  The resamples originally come from the tsml java library.  From memory, the tsml library doesn't have any way to "export" the resamples, so we wrote some code (a long time ago) to extract the resample indices for most but maybe not all of the datasets.  Those are the indices in the multirocket github.  These indices did match the resamples generated by the tsml library, at least at that time.

If I remember correctly, the latest papers from the UEA group, e.g., the new "bake off" paper, use a mixture of different resamples: the "old" 112 datasets are using resamples generated from the java library, and the "new" 30 datasets are using resamples generated from python.

I would expect the indices in the multirocket library to match the bakeoff resamples for the "old" 112 datasets.  If not, maybe there has been some changes over time in the java library which means that the random state is producing different results.  I don't know, it's not something I can control.

Obviously, java and python will produce completely different resamples as they use totally separate random number generators etc.

Have you had a look here?

<https://github.com/time-series-machine-learning/tsml-eval/blob/main/tsml_eval/publications/y2023/tsc_bakeoff/tsc_bakeoff_2023.ipynb>

It looks like you can download the full resampled datasets, that way you would at least be fairly confident that your results should match the new "bake off" paper.  You could also contact Matthew Middlehurst through the tsml-eval or aeon githubs if you have any questions or want to check something about the resamples directly with the UEA team.

For the "new" 30 datasets, I made the resamples using the `stratified_resample_data` function from `tsml_eval/utils/experiments.py` <https://github.com/time-series-machine-learning/tsml-eval/blob/5dac9ce9955bfb4596735ef117438c4c2a299acc/tsml_eval/utils/experiments.py#L98>.

I will check to see whether the multirocket indices match the current resamples at least for the "old" 112 datasets.

Hopefully over time all the resamples are somehow made clear and easy and all in one place.

I hope that helps a little bit.  Sorry I can't be of more help.

Best,

Angus

**Comment by firrice** [comment]
@angus924 Really thanks for your detailed explanation! Now i'm clear about the situation of resample indices, and i also wish resample indices from python&java version can be unified in the future.

**Comment by angus924** [comment]
Hi @firrice,

I've double checked the multirocket github indices against the full resampled files available through:

https://github.com/time-series-machine-learning/tsml-eval/blob/main/tsml_eval/publications/y2023/tsc_bakeoff/tsc_bakeoff_2023.ipynb

...they do indeed match (subject to some numerical differences).  That means you should be able to use the multirocket github indices to reproduce the resamples for the "old" ucr datasets.

Best,

Angus

**Comment by firrice** [comment]
@angus924 Sorry for my slow reply. I'm still downloading the resamples datasets(8.7GB) from the link(https://github.com/time-series-machine-learning/tsml-eval/blob/main/tsml_eval/publications/y2023/tsc_bakeoff/tsc_bakeoff_2023.ipynb) , it's slow because of my own network factor. And actually i just found resample indices of UCR109,  not UCR112 in MultiRocket's github(https://github.com/ChangWeiTan/MultiRocket/blob/main/data/). 

Here i list the indices of 10 sequences contained the first category of the ACSF1 train_dataset(resample_id=1) based on tsml-eval、tsml-java and MultiRocket.

(1) MultiRocket indices : 32, 137, 36, 135, 131, 133, 139, 38, 37, 132 (reference to: https://github.com/ChangWeiTan/MultiRocket/blob/main/data/indices109/ACSF1_INDICES_TRAIN.txt)

(2)tsml-eval indices: 33, 136, 36, 130, 32, 134, 34, 137, 37, 31(which used stratified_resample_data() of tsml-eval and the results are same in version0.0.1 and version 0.1.1 of tsml-eval)

(3)tsml-java indices: 33, 138, 37, 136, ...(tsml-java don't offer code to generate indices, so i can only dope out it through observing the value of resample sequences generated by tsml-java)

Summarized as follows:
(1) 32, 137, 36, 135, 131, 133, 139, 38, 37, 132
(2) 33, 136, 36, 130, 32, 134, 34, 137, 37, 31
(3) 33, 138, 37, 136, ...
Indices above are different with each other and i will check it when i have downloaded the UCR112 resample datasets. :-)

**Comment by firrice** [comment]
@angus924 Hi, i have downloaded the UCR112 resample datasets from https://github.com/time-series-machine-learning/tsml-eval/blob/main/tsml_eval/publications/y2023/tsc_bakeoff/tsc_bakeoff_2023.ipynb, and i have checked it. I think i get  clear about it, that is the UCR112 resample datasets do indeed match the results of MultiRocket. 

Here is my analysis which was based on 10 sequences of the first category(that is label '0') of the ACSF1 train dataset, resample_id=1:

(1) tsc_bakeoff / ACSF11_TRAIN.ts(resample_id=1): 
![image](https://github.com/angus924/quant/assets/26590498/9cd3f73b-d92e-47e8-8d2b-ca29f99f78f0)
we can find values(in the red box of the fig above) in the ACSF1_TRAIN.ts(resample_id=0) to dope out indices respectively, the result indices are:
32, 137, 36, 135

Here, i summarize it again(based on ACSF1 train dataset, resample id=1, which id belong to {0,1,2,...,29}):
(1) MultiRocket: 32, 137, 36, 135, 131, 133, 139, 38, 37, 132
(2) tsml-eval(v0.1.0&v0.1.1): 33, 136, 36, 130, 32, 134, 34, 137, 37, 31
(3) tsml-java(current version): 32, 137, 36, 135, ...(Not "33, 138, 37, 136, ..." in my older answer, sorry i mistaked starting it from 1 rather than 0)
(4) tsc_bakeoff: 32, 137, 36, 135, ...

Conlusion(to prevent others from making similar mistakes):  tsc_bakeoff、MultiRocket and tsml-java use the sample resamples and tsml-eval use its own.

Finally, really thanks so much for every detailed and quick reply! Best wishes for u! :-)
